internal/sourcer/source.go

```diff
@@ -0,0 +1,89 @@
+package sourcer
+
+import (
+	"context"
+
+	"github.com/blang/semver/v4"
+	registryClient "github.com/operator-framework/operator-registry/pkg/client"
+
+	v1alpha1 "github.com/timflannagan/platform-operators/api/v1alpha1"
+)
+
+type Sourcer interface {
+	Source(context.Context, *v1alpha1.PlatformOperator) (*Bundle, error)
+}
+
+type Bundle struct {
+	Version  string
+	Image    string
+	Replaces string
+	Skips    []string
+}
+
+type catalogSource struct {
+	client registryClient.Interface
+}
+
+func NewCatalogSourceHandler(c registryClient.Interface) Sourcer {
+	return &catalogSource{
+		client: c,
+	}
+}
+
+func (cs catalogSource) Source(ctx context.Context, po *v1alpha1.PlatformOperator) (*Bundle, error) {
+	bundles, err := getBundles(ctx, cs.client, po)
+	if err != nil {
+		return nil, err
+	}
+	filteredBundle, err := filterBundles(bundles, func(b1, b2 *Bundle) bool {
```

Comment:
Could we add a `len(bundles)==0` check here and return early in that case? Otherwise I believe we end up with a nil pointer as the desiredBundle